Sema: fix analysis of inline calls of generic fns

castholm opened the pull request:
Fixes comptime-known arguments not being able to be resolved when evaluating inline calls of generic functions, which could result in unintentional compile errors if a generic parameter or return type depended on the parameter value.

Resolves: #22494
Resolves: #22532

cc @mlugg

mlugg commented:
In the test, please store the function return values and:
* `comptime assert` the return values' types are what's expected
* `try expect` the return values are what's expected

Also, you could simplify the test a little by making the generic return types just `Something(the_param)` (without the `@TypeOf`), but what you have now is fine too.

castholm commented:
@mlugg I updated the test to include a generic parameter that depends on a prior parameter (I verified that this is indeed also a compile error on master) and changed the test name and rewrote the commit/PR description to reflect this.

After reading the Zulip discussion I *think* I understand why this PR didn't fix #22538: it's because those functions are not considered generic under the current language rules. If we make them generic (e.g. by making one of the parameters `anytype`) we get the intended results.

Feel free to close/supercede this PR if it gets in the way of you taking a look at #22538. But if you do end up including this test in your changeset, I believe you could change the types of the first parameters to `u32` to ensure that the new proposed "fn is generic if a parameter/the return type attempts to evaluate a prior parameter" rules are tested:

```diff
diff --git a/test/behavior/call.zig b/test/behavior/call.zig
index e00a441eb5..f0b6e7246b 100644
--- a/test/behavior/call.zig
+++ b/test/behavior/call.zig
@@ -758,10 +758,10 @@ test "tail call function pointer" {
 
 test "inline call of generic function with parameter and return type that evaluate prior parameter" {
     const S = struct {
-        inline fn foo1(x: anytype, y: Bar(x)) Bar(x) {
+        inline fn foo1(x: u32, y: Bar(x)) Bar(x) {
             return x + y;
         }
-        fn foo2(x: anytype, y: Bar(x)) Bar(x) {
+        fn foo2(x: u32, y: Bar(x)) Bar(x) {
             return x + y;
         }
         fn Bar(comptime x: comptime_int) type {
```

mlugg commented:
The issue with generics is a *little* more subtle than your explanation, but you've got the gist of it. I'll probably ultimately supercede this PR, but thanks anyway!
